File: rust/src/ml_gol.rs
use crate::game_of_life::Gol;
use crate::color::{Color, ColorHSV};
use crate::image::write_png_file;

use rayon::prelude::*;

pub struct MlGol {
    grid_size: usize,
    layers: Vec<Gol>,
    combined: Vec<Color>,
    dependent: Vec<Color>
}

impl MlGol {
    pub fn new(grid_size: u32, num_layers: u32, density: f32) -> MlGol {
        let mut layers = Vec::new();
        let mut combined = Vec::new();
        let mut dependent = Vec::new();

        // Add 2 to the grid size to account for the ghost cells
        let grid_size = (grid_size + 2) as usize; 
        let num_layers = num_layers as usize;

        for _ in 0..num_layers {
            layers.push(Gol::new(grid_size, density));
        }

        // Initialize the combined and dependent grids with black color
        for _ in 0..grid_size * grid_size {
            combined.push(Color { r: 0, g: 0, b: 0 });
            dependent.push(Color { r: 0, g: 0, b: 0 });
        }

        println!("Initialized multilayer game of life with {} layers and grid size {}", num_layers, grid_size - 2);
        MlGol::print_layers_colors(layers.len());

        MlGol {
            grid_size,
            layers,
            combined,
            dependent
        }
    }

    fn print_layers_colors(num_layers: usize) {
        println!("Colors for the layers:");
        for i in 0..num_layers {
            let color = MlGol::get_color_for_layer(i, num_layers);
            println!("Layer {}: {}", i, Color::color_to_hex(&color));
        }
    }

    pub fn start_game(&mut self, num_steps: usize) {
        for step in 0..num_steps {
            for layer in self.layers.iter_mut() {
                layer.fill_ghost_cells();
                layer.step();    
                layer.swap_grids();   
            }

            self.calculate_combined();
            self.calculate_dependent();

            self.create_png_for_step(step);
            self.reset_combined_and_dependent();
        }
    }

    pub fn start_game_par(&mut self, num_steps: usize) {
        for step in 0..num_steps {
            for layer in self.layers.iter_mut() {
                layer.fill_ghost_cells();
                layer.step_par();    
                layer.swap_grids();   
            }

            self.calculate_combined_par();
            self.calculate_dependent_par();

            self.create_png_for_step_par(step);
            self.reset_combined_and_dependent_par();
        }
    }

    fn get_color_for_layer(layer: usize, num_layers: usize) -> Color {
        // Calculate the angle for the hue based on the layer
        let hue = (360.0 * layer as f64) / num_layers as f64;

        let hsv_color = ColorHSV {
            h: hue,
            s: 1.0,
            v: 1.0,
        };

        hsv_color.to_rgb()
    }

    fn calculate_combined(&mut self) {

        for k in 0..self.layers.len() {
            let color = MlGol::get_color_for_layer(k, self.layers.len());

            for i in 0..self.grid_size {
                for j in 0..self.grid_size {
                    let idx = (i * self.grid_size + j) as usize;

                    if !self.layers[k].current[idx] {
                        continue;
                    }

                    self.combined[idx] = Color::add_colors(&self.combined[idx], &color);
                }
            }
        }
    }

    fn calculate_combined_par(&mut self) {
        self.combined = (0..self.combined.len())
            .into_par_iter()
            .map(|idx| {
                let point = Gol::point(self.grid_size, idx);

                if point.x == 0 || point.x == self.grid_size - 1 || point.y == 0 || point.y == self.grid_size - 1 {
                    return Color { r: 0, g: 0, b: 0 };
                }

                let mut color = Color { r: 0, g: 0, b: 0 };

                for k in 0..self.layers.len() {
                    let layer = &self.layers[k];
                    let layer_color = MlGol::get_color_for_layer(k, self.layers.len());

                    if layer.current[idx] {
                        color = Color::add_colors(&color, &layer_color);
                    }
                }

                color
            })
            .collect();
    }

    fn create_png_for_grid(grid: &Vec<Color>, grid_size: usize, step: usize, folder: &str) {
        let mut buffer = vec![0u8; (grid_size * grid_size * 4) as usize];

        for i in 0..grid_size {
            for j in 0..grid_size {
                let idx = ((i * grid_size + j) * 4) as usize;
                let grid_idx = (i * grid_size + j) as usize;

                buffer[idx] = grid[grid_idx].r;
                buffer[idx + 1] = grid[grid_idx].g;
                buffer[idx + 2] = grid[grid_idx].b;
                buffer[idx + 3] = 255;
            }
        }

        let filename = format!("output/{}/{}{:04}.png", folder, folder, step);
        write_png_file(&filename, grid_size as u32, grid_size as u32, &buffer);
    }

    fn create_png_for_grid_par(grid: &Vec<Color>, grid_size: usize, step: usize, folder: &str) {
        let buffer: Vec<u8> = (0..grid_size * grid_size)
            .into_par_iter()
            .flat_map(|idx| {
                let i = idx / grid_size;
                let j = idx % grid_size;

                let grid_idx = (i * grid_size + j) as usize;

                vec![grid[grid_idx].r, grid[grid_idx].g, grid[grid_idx].b, 255]
            })
            .collect();

        let filename = format!("output/{}/{}{:04}.png", folder, folder, step);
        write_png_file(&filename, grid_size as u32, grid_size as u32, &buffer);
    }

    fn create_png_for_step(&mut self, step: usize) {
        let grid_size = self.grid_size - 2;

        MlGol::create_png_for_grid(&self.combined, grid_size, step, "combined");
        MlGol::create_png_for_grid(&self.dependent, grid_size, step, "dependent");
    }

    fn create_png_for_step_par(&mut self, step: usize) {
        let grid_size = self.grid_size - 2;

        MlGol::create_png_for_grid_par(&self.combined, grid_size, step, "combined");
        MlGol::create_png_for_grid_par(&self.dependent, grid_size, step, "dependent");
    }

    fn reset_combined_and_dependent(&mut self) {
        for i in 0..self.combined.len() {
            self.combined[i] = Color { r: 0, g: 0, b: 0 };
            self.dependent[i] = Color { r: 0, g: 0, b: 0 };
        }
    }

    fn reset_combined_and_dependent_par(&mut self) {
        self.combined = (0..self.combined.len())
            .into_par_iter()
            .map(|_| Color { r: 0, g: 0, b: 0 })
            .collect();

        self.dependent = (0..self.dependent.len())
            .into_par_iter()
            .map(|_| Color { r: 0, g: 0, b: 0 })
            .collect();
    }

    fn count_dependent_alive_neighbors(&self, i: usize, j: usize) -> u8 {
        let mut count = 0;

        for x in -1..=1 {
            for y in -1..=1 {
                if x == 0 && y == 0 {
                    continue;
                }

                let idx = ((i as i32 + x) * self.grid_size as i32 + (j as i32 + y)) as usize;

                if self.combined[idx].r > 0 && self.combined[idx].g > 0 && self.combined[idx].b > 0 {
                    count += 1;
                }
            }
        }

        count
    }

    fn calculate_dependent(&mut self) {
        for i in 1..self.grid_size - 1 {
            for j in 1..self.grid_size - 1 {
                let idx = (i * self.grid_size + j) as usize;
                
                let alive_neighbors = self.count_dependent_alive_neighbors(i, j);

                let channel_value = ((alive_neighbors as f32) / 8.0 * 255.0) as u8;
                
                self.dependent[idx] = Color { r: channel_value, g: channel_value, b: channel_value };
                
            }
        }
    }

    fn calculate_dependent_par(&mut self) {
        self.dependent = (0..self.dependent.len())
            .into_par_iter()
            .map(|idx| {
                let point = Gol::point(self.grid_size, idx);

                if point.x == 0 || point.x == self.grid_size - 1 || point.y == 0 || point.y == self.grid_size - 1 {
                    return Color { r: 0, g: 0, b: 0 };
                }

                let alive_neighbors = self.count_dependent_alive_neighbors(point.x, point.y);
                let channel_value = ((alive_neighbors as f32) / 8.0 * 255.0) as u8;

                Color { r: channel_value, g: channel_value, b: channel_value }
            })
            .collect();
    }


}

File: openmp/src/ml_gol.c
#include "ml_gol.h"

#include <stdlib.h>
#include <stdio.h>
#include <omp.h>

void start_game(const uint64_t grid_size, const uint64_t num_layers, const uint64_t num_steps, const bool create_png, const float density, const uint64_t seed) {
    ml_gol_t* ml_gol = (ml_gol_t*) malloc(sizeof(ml_gol_t));

    init_ml_gol(ml_gol, grid_size, num_layers, create_png, density, seed);

    printf("Starting simulation with %ld steps and %d threads\n", num_steps, omp_get_num_threads());

    for (uint64_t s = 1; s < num_steps; s++) {
#pragma omp parallel for
        for (uint64_t layer = 0; layer < num_layers; layer++) {
            step(&ml_gol->layers[layer]);
        }
       
        calculate_combined(ml_gol); 
        calculate_dependent(ml_gol);

        if (create_png) {
            create_png_for_step(ml_gol, s);
        }

        reset_combined_and_dependent(ml_gol);
    }
    
    free_ml_gol(ml_gol);
}


void calculate_combined(const ml_gol_t* ml_gol) {
    for (uint64_t layer = 0; layer < ml_gol->num_layers; layer++) {
#pragma omp parallel for collapse(2)
        for (uint64_t i = 1; i < ml_gol->grid_size + 1; i++) {
            for (uint64_t j = 1; j < ml_gol->grid_size + 1; j++) {
                size_t grid_idx = idx(&ml_gol->layers[layer], i, j);
                size_t combined_idx = (i - 1) * ml_gol->grid_size + (j - 1);
                
                // If the cell is alive in the current layer, add the color of the layer to the combined grid
                if (ml_gol->layers[layer].current[grid_idx]) {
                    ml_gol->combined[combined_idx] = add_colors(ml_gol->combined[combined_idx], ml_gol->layers_colors[layer]);
                }
            }
        }
    }
}


void create_png_for_grid(const color_t* grid, const uint64_t grid_size, const uint64_t step, const char* folder) {
    char filename[50];

    // 3 channels: RGB
    const uint8_t channels = 3;
    uint8_t* buffer = (uint8_t*) malloc(grid_size * grid_size * channels * sizeof(uint8_t));

#pragma omp parallel for collapse(2)
    for (uint64_t i = 0; i < grid_size; i++) {
        for (uint64_t j = 0; j < grid_size; j++) {
            uint64_t idx = (i * grid_size + j);

            buffer[idx * channels] =     grid[idx].r;
            buffer[idx * channels + 1] = grid[idx].g;
            buffer[idx * channels + 2] = grid[idx].b;
        }
    }

    sprintf(filename, "output/%s/%s%04ld.png", folder, folder, step);
    write_png_file(filename, grid_size, grid_size, buffer);

    free(buffer);
}

void create_png_for_step(const ml_gol_t* ml_gol, const uint64_t step) {
    create_png_for_grid(ml_gol->combined, ml_gol->grid_size, step, "combined");
    create_png_for_grid(ml_gol->dependent, ml_gol->grid_size, step, "dependent");
}

void reset_combined_and_dependent(ml_gol_t* ml_gol) {
#pragma omp parallel for
    for (uint64_t i = 0; i < ml_gol->grid_size * ml_gol->grid_size; i++) {
        ml_gol->combined[i] = BLACK;
        ml_gol->dependent[i] = BLACK;
    }
}

void print_layers_colors(const ml_gol_t* ml_gol) {
    char hex[8];
    printf("Colors for the layers:\n");
    for (uint64_t i = 0; i < ml_gol->num_layers; i++) {
        color_to_hex(ml_gol->layers_colors[i], hex);
        printf("Layer %ld: %s\n", i, hex);
    }
}

void init_ml_gol(ml_gol_t* ml_gol, const uint64_t grid_size, const uint64_t num_layers, const bool create_png, const float density, const uint64_t seed) {
    srand(seed);

    ml_gol->num_layers = num_layers;
    ml_gol->layers = (gol_t*) malloc(num_layers * sizeof(gol_t));
    ml_gol->layers_colors = (color_t*) malloc(num_layers * sizeof(color_t));

    ml_gol->grid_size = grid_size;

    for (uint64_t i = 0; i < ml_gol->num_layers; i++) {
        init_gol(&ml_gol->layers[i], grid_size, density);
        ml_gol->layers_colors[i] = get_color_for_layer(i, num_layers);
    }

    size_t size = (ml_gol->grid_size) * (ml_gol->grid_size) * sizeof(color_t);

    ml_gol->combined = (color_t*) malloc(size * sizeof(color_t));
    ml_gol->dependent = (color_t*) malloc(size * sizeof(color_t));

    calculate_combined(ml_gol); 
    calculate_dependent(ml_gol);

    if (create_png) {
        create_png_for_step(ml_gol, 0);
    }

    reset_combined_and_dependent(ml_gol);

    printf("Initialized multilayer game of life with %ld layers and grid size %ld\n", num_layers, grid_size);
    
    print_layers_colors(ml_gol);
}

color_t get_color_for_layer(const uint64_t layer, const uint64_t num_layers) {
    // Calculate the angle for the hue based on the layer
    double hue = (double) layer / num_layers * 360.0;

    color_hsv_t hsv_color;
    hsv_color.h = hue;
    hsv_color.s = 1.0; 
    hsv_color.v = 1.0;

    return hsv_to_rgb(hsv_color);
}

uint8_t count_dependent_alive_neighbors(const ml_gol_t* ml_gol, const uint64_t i, const uint64_t j) {
    uint8_t count = 0;

    for (int32_t x = -1; x <= 1; x++) {
        for (int32_t y = -1; y <= 1; y++) {
            uint64_t layer_idx = idx(&ml_gol->layers[0], i + x, j + y);

            for (uint64_t layer = 0; layer < ml_gol->num_layers; layer++) {
                if (ml_gol->layers[layer].current[layer_idx]) {
                    count++;
                }
            }
        }
    }

    return count;
}

void calculate_dependent(const ml_gol_t* ml_gol) {
#pragma omp parallel for collapse(2)
    for (uint64_t i = 1; i < ml_gol->grid_size - 1; i++) {
        for (uint64_t j = 1; j < ml_gol->grid_size - 1; j++) {
            uint64_t idx = i * ml_gol->grid_size + j;
            
            uint8_t alive_neighbors = count_dependent_alive_neighbors(ml_gol, i, j);

            uint8_t channel_value = (uint8_t) ((((float) alive_neighbors) / (9 * ml_gol->num_layers)) * 255);
            
            ml_gol->dependent[idx] = (color_t){channel_value, channel_value, channel_value};
            
        }
    }
}

void free_ml_gol(ml_gol_t* ml_gol) {
    for (uint64_t i = 0; i < ml_gol->num_layers; i++) {
        free_gol(&ml_gol->layers[i]);
    }

    free(ml_gol->layers);
    free(ml_gol->combined);
    free(ml_gol->dependent);
    free(ml_gol);
}